### PR TITLE
New rule S6669: Logger field names should comply with a naming convention

### LINKED ISSUE #8842: New rule S6669: Logger field names should comply with a naming convention
See https://github.com/SonarSource/rspec/pull/2533

### PR DESCRIPTION
**Pull request by gregory-paidis-sonarsource**
Fixes #8842

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs:59]: Looking at the rules of MMF#2, I think we _really_ need to prioritize https://github.com/SonarSource/sonar-dotnet/issues/8851, since a lot of them need to check every single (field|invocation|argument), which will probably be more expensive than checking if a library is referenced.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs:32]: Looking at the other parametrized analyzers, we pass `isEnabledByDefault` to false.
Do you know why this is the case?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs]: We could throw it to peach and see how bad the performance is, but I still think we should implement a library-reference check.

**Comment by costin-zaharia-sonarsource** [comment]
Before moving on with this PR I think we need to have some performance checks to see the impact of the rule.

File: analyzers/tests/SonarAnalyzer.Test/Rules/LoggerMembersNamesShouldComplyTest.cs
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/Rules/LoggerMembersNamesShouldComplyTest.cs:80]: pedantic

```suggestion
        Builder.AddSnippet("""
```

File: analyzers/tests/SonarAnalyzer.Test/Rules/LoggerMembersNamesShouldComplyTest.cs
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/Rules/LoggerMembersNamesShouldComplyTest.cs:103]: pedantic

```suggestion
        Builder.AddSnippet("""
```

Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/Rules/LoggerMembersNamesShouldComplyTest.cs:103]: same for the rest

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]
I think the implementation looks good, there are just a few details to tackle.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs:32]: Thanks a lot, very useful info!

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs:40]: Yeah, very good point.
The problem is you cannot make it work like this, which creates a backing field:

```csharp
    private Regex NameRegex { get; } = new(Format, RegexOptions.Compiled, RegexConstants.DefaultTimeout);
```

...because of the `Format` reference, which is a property.

I initialize it on the `CompilationStart` action to do it only once per compilation.

Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs:40]: Another option would be to use (or be 😛) [Lazy<T>](https://learn.microsoft.com/en-us/dotnet/api/system.lazy-1?view=net-8.0)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs:58]: I really like it!
Furthermore, to optimize the check "has the user changed the regex", I chose to set the default format to an empty string. 
Let me know if you find this confusing.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs:68]: You can rebase on the other PR and remove the ToDo. Up to you.

File: analyzers/tests/SonarAnalyzer.Test/Rules/LoggerMembersNamesShouldComplyTest.cs
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/Rules/LoggerMembersNamesShouldComplyTest.cs:180]: typo: `parameterized`

File: analyzers/tests/SonarAnalyzer.Test/Rules/LoggerMembersNamesShouldComplyTest.cs
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/Rules/LoggerMembersNamesShouldComplyTest.cs:184]: pedantic: indentation is a bit off

**Comment by costin-zaharia-sonarsource** [comment]
I think it looks good, I'm waiting for the ToDo to be removed and for the perf stats.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerMembersNamesShouldComply.cs:68]: That's my plan, I left the TODO to break the QG and not merge this until we merge the helper PR and I can use the `StartCompilation` check :)

File: analyzers/tests/SonarAnalyzer.Test/Rules/LoggerMembersNamesShouldComplyTest.cs
Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/Rules/LoggerMembersNamesShouldComplyTest.cs:184]: still is :). Sorry, I was not clear, all the method invocations should be shifted right with 4 spaces.

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**
